### PR TITLE
kubeadm: update docker version for CE and EE

### PR DESCRIPTION
**Pull request by dmmcquay**
**What this PR does / why we need it**: Update regex for docker version to also capture new CE and EE versions. 

**Which issue this PR fixes**: fixes #https://github.com/kubernetes/kubeadm/issues/189

**Special notes for your reviewer**: /cc @jbeda @luxas

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by luxas** [comment]
cc @Random-Liu @dchen1107 @vishh @yujuhong @kubernetes/sig-node-pr-reviews 
Is docker 1.13.x validated and officially supported in v1.6?

**Comment by yujuhong** [comment]
> Is docker 1.13.x validated and officially supported in v1.6?

Nope. It's not running in node e2e yet.

**Comment by jbeda** [comment]
Discussion right now if we should allow 1.9+ without qualification (as this PR as it stands) or if we should warn for 1.13+ as it hasn't been 100% validated yet.

**Comment by luxas** [comment]
> Discussion right now if we should allow 1.9+ without qualification (as this PR as it stands) or if we should warn for 1.13+ as it hasn't been 100% validated yet.

I think we should absolutely not allow 1.9+ without warnings; that would be bad UX.

If docker 1.13 isn't validated for v1.6, it isn't

I think we should continue to fail for 1.13 and 17.03-ce until sig-node has validated it.

On the other hand, our docs could include that "if you know what you're doing and have run this iptables command described in issue xxx then you may move on with `--skip-preflight-checks`"

File: test/e2e_node/system/types.go [hunk 1]
```diff
@@ -114,7 +114,7 @@ var DefaultSysSpec = SysSpec{
 	Cgroups: []string{"cpu", "cpuacct", "cpuset", "devices", "freezer", "memory"},
 	RuntimeSpec: RuntimeSpec{
 		DockerSpec: &DockerSpec{
-			Version: []string{`1\.(9|\d{2,})\..*`}, // Requires 1.9+
+			Version: []string{`1\.(9|\d?[0-2]{2,})\..*`}, // Requires 1.9+
```
Comment (Random-Liu) [inline test/e2e_node/system/types.go]: I'm not sure I understand the regexp here.

If we only want to officially support 1.9 - 1.12, could we change this to `1\.(9|1[0-2])\..*`?

@dchen1107 Two questions here:
1) Do we still support 1.9?
2) Are we going to say we don't support 1.13+ now?

**Comment by dmmcquay** [comment]
@k8s-bot cvm gce e2e test this
@k8s-bot kops aws e2e test this

File: test/e2e_node/system/types.go (hunk 1 above)
Comment (vishh) [inline test/e2e_node/system/types.go]: Testing code could still allow for testing newer versions. We might get some early signals from the community before upstream validates new versions. WDYT @Random-Liu ?

**Comment by dmmcquay** [comment]
@k8s-bot cvm gce e2e test this

**Comment by jbeda** [comment]
This LGTM.  The only thing I'd like to see is some pointer for the user to learn more.  Based on our current docs I think that https://kubernetes.io/docs/getting-started-guides/kubeadm/#limitations is the best place to send people.

I'm thinking we can just do this as part of the mondo error returned by `RunChecks` in `checks.go`.  If we don't have something here it is likely that we'll have a lot of support issues to deal with.

I'm ok with merging without that, but I'd feel better if we had it.

File: test/e2e_node/system/types.go (hunk 1 above)
Comment (Random-Liu) [inline test/e2e_node/system/types.go]: > Testing code could still allow for testing newer versions. We might get some early signals from the community before upstream validates new versions. WDYT @Random-Liu ?

Good point, sounds good to me.

However, for kubeadm preflight check, should we fail on version other than 1.10-1.12?

File: test/e2e_node/system/docker_validator_test.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ func TestValidateDockerInfo(t *testing.T) {
 		Reporter: DefaultReporter,
 	}
 	spec := &DockerSpec{
-		Version:     []string{`1\.(9|\d{2,})\..*`},
+		Version:     []string{`1\.(9|\d?[0-2]{2,})\..*`}, // Requires 1.9+
```
Comment (Random-Liu) [inline test/e2e_node/system/docker_validator_test.go]: The regex seems wrong to me. For example things like `1.21`, `1.221` will also match.

If we only want to support 1.9-1.12, `1\.(9|1[0-2])\..*` seems better to me

Comment (dchen1107) [inline test/e2e_node/system/docker_validator_test.go]: sig-node agreed to drop 1.9.x support for 1.6 release, can we update the test here reflecting such change too?

**Comment by dchen1107** [comment]
Can we change the docker version checking logic here:

1) If docker version < Kubelet's docker minimal support version (docker client: 1.10, docker server 1.22, for k8s 1.6), throwing an error;
2) If docker version in the range of being validated and supported by the node team (docker 1.10.x to docker 1.12.x for k8s 1.6), pass;
3) if docker version is greater than the maximum version validated by the node team (for example: docker 1.13, etc.), throwing a warning with the proper validation information
So that node-e2e tests can still run against latest docker release. Of course, docker changed their version schema, we are going to address it in 1.7 timeframe.

WDYT?

**Comment by jbeda** [comment]
I'm okay with @dchen1107's proposal.  Currently the validators don't have a way to return warnings so that'll require some new plumbing.  

@dmmcquay -- you up for it?

I still think we should direct users to a page for any precheck failure.  At least we can address common things quickly there without pushing new code.

**Comment by yujuhong** [comment]
> Of course, docker changed their version schema, we are going to address it in 1.7 timeframe.

We should check the docker API version (which is semver) instead.

**Comment by dmmcquay** [comment]
@dchen1107 @jbeda I don't know if I agree with the proposed change. Is the idea to rework the `Validator interface` so that `Validate` returns both a warning and an error?  It seems like changing the other validators to compensate for just verifying docker would be strange. 

What is the purpose of throwing the "warning"? Would logging the info satisfy this requirement?

**Comment by jbeda** [comment]
@yujuhong -- The API version isn't enough.  There are other changes (such as how Docker sets up iptables) that can break people and won't show up as a breaking change to the API.

@dmmcquay -- my feeling is that <=1.9 we know it *wont* work.  For 1.10-1.12 we have tested and validated.  For >=1.13 it hasn't been validated and there may be issues (known and not known).  

If we error for 1.13+ then if there are issues that people can work around they'll have to turn off all preflight checks.  That means they may miss other problems.  If we were able to specify which prechecks to skip then I'd be more comfortable making it an error over a warning.

This would mean changing the validator interface and that would be a bigger change.  But the preflight check interface already knows how to surface warnings so it'll be limited to just the validator interface.

Thoughts?  @dchen1107 @luxas @lukemarsden?

If we can't come to some other conclusion by EOD today, I think we should go with error for all non-proven versions with the addition of a link to the kubeadm docs.

**Comment by dmmcquay** [comment]
@jbeda One solution that I thought of would be making a new error type:

```go
 type DockerValidatorError struct {  
     Msg     string                   
     Warning string                   
 }                                    
```

And then do some sort of type switch on the error, and if its a `DockerValidatorError`, you can check to see if there warning string is empty or not. 

Not saying this is a better solution, just an idea for keeping the interface intact.

**Comment by jbeda** [comment]
@dmmcquay I dunno.  That just seems to move the problem around into the kubeadm code.  At the end of the day, we should look to refactor this stuff out of a test directory so we shouldn't be afraid to move it forward (as long as @dchen1107 and @yujuhong are cool with it).

**Comment by dmmcquay** [comment]
@jbeda, @dchen1107, @yujuhong PTAL and let me know what you think about the changes I made. This not will allow 1.13+ (including EE and CE versions) but will return a warning saying `docker version is greater than the maximum version validated by the node team`.

**Comment by dmmcquay** [comment]
@k8s-bot non-cri e2e test this

File: test/e2e_node/system/docker_validator.go [hunk 3]
```diff
@@ -70,17 +70,29 @@ func (d *DockerValidator) validateDockerInfo(spec *DockerSpec, info types.Info)
...
 	}
 	if !matched {
+		// catch if docker is 1.13+
+		ver := `1\.(1[3-9])\..*|\d{2}\.\d+\.\d+-[a-z]{2}`
+		r := regexp.MustCompile(ver)
+		if r.MatchString(info.ServerVersion) {
+			fmt.Println("getting here")
+			d.Reporter.Report(dockerConfigPrefix+"VERSION", info.ServerVersion, good)
+			w := fmt.Errorf(
+				"docker version is greater than the maximum version validated by the node team. docker version: %s",
```
Comment (jbeda) [inline test/e2e_node/system/docker_validator.go]: Users running this won't know who the "node team" is.  I'd do something like:

"docker version is greater than the most recently validated version. Docker version: %s. Max validated version: %s"

Ideally we could point to a URL about what the known issues might be.  @dchen1107 @yujuhong  -- does the node team keep any documentation about the latest status of what has been validated?

File: test/e2e_node/system/docker_validator_test.go (hunk 2 above)
Comment (jbeda) [inline test/e2e_node/system/docker_validator_test.go]: We should update this now.

**Review by jbeda — CHANGES_REQUESTED** [review]
Just some small nits.  This didn't end up being as invasive as we feared!

File: test/e2e_node/system/docker_validator.go [hunk 4]
```diff
@@ -70,17 +70,29 @@ func (d *DockerValidator) validateDockerInfo(spec *DockerSpec, info types.Info)
 		}
 	}
 	if !matched {
+		// catch if docker is 1.13+
+		ver := `1\.(1[3-9])\..*|\d{2}\.\d+\.\d+-[a-z]{2}`
+		r := regexp.MustCompile(ver)
+		if r.MatchString(info.ServerVersion) {
+			fmt.Println("getting here")
```
Comment (luxas) [inline test/e2e_node/system/docker_validator.go]: nit: remove

**Review by jbeda — APPROVED** [review]

**Comment by dmmcquay** [comment]
@k8s-bot non-cri e2e test this

**Comment by luxas** [comment]
Applying @jbeda's LGTM...

**Pull request merged**